### ISSUE TITLE
Create custom dataset class for template-agnostic fine-tuning

### ISSUE THREAD

**Issue by niznik-dev** [niznik-dev/predicting-zygosity#241]
## Motivation

Our investigation into fine-tuning failures revealed that torchtune's chat templates and HuggingFace's `apply_chat_template()` handle special tokens differently. This creates training/inference mismatches that break learned patterns:

- **Header tokens** - `<|start_header_id|>`, `<|end_header_id|>` placement varies
- **Turn tokens** - `<|eot_id|>` injection differs between frameworks
- **Preamble tokens** - System prompt preprocessing (dates, metadata) added inconsistently
- **BOS/EOS handling** - When and whether these get added is framework-dependent

When positional embeddings shift by even a few tokens, fine-tuned behavior breaks down. We're currently at the mercy of opaque template implementations.

## Proposed Solution

Create a `ConditionalCompletionDataset` class that:

1. **Bypasses chat templates entirely** - Uses raw tokenization with explicit special token handling
2. **Configurable prompt format** - String templates or callables for flexibility
3. **Explicit masking control** - `target_only` masks prompt tokens with -100, `full` trains on everything
4. **Separator control** - Explicit control over what goes between prompt and target
5. **Tokenization boundary verification** - Raises error if prefix tokenization doesn't match

## Key Features

```python
@dataclass
class ConditionalCompletionConfig:
    prompt: PromptSpec          # "{input}" or callable
    input_key: str = "input"
    output_key: str = "output"
    add_bos: bool = True
    add_eos: bool = True
    max_length: Optional[int] = None
    mask: str = "target_only"   # "target_only" | "full"
    separator: str = ""
```

## inspect-ai Integration

For short-answer tasks (like capitalization, token reversal), even a single token mismatch between training and inference breaks evaluation. We need a corresponding approach for inspect-ai evaluation that:

- Uses the same prompt construction logic as training
- Avoids `apply_chat_template()` or uses identically-configured tokenizer
- Generates prompts directly rather than relying on inspect's message formatting

This might involve a custom `generate()` solver or a thin wrapper that ensures the model sees the exact same token sequence at eval time as it did during training.

## Benefits

- **Training/inference parity** - Same tokenization logic for both
- **No framework surprises** - We control every token
- **Debuggable** - Can inspect exact token sequences
- **Task-agnostic** - Works for any conditional completion task

## Implementation Notes

- Needs integration with torchtune's training loop (custom recipe or dataset wrapper)
- Include `pad_collate` function for batching
- Should work with existing `setup_finetune.py` workflow
- Corresponding inspect-ai solver/wrapper for evaluation parity
